### ISSUE TITLE
Integrate vector databases with `RetrievalResults`

### ISSUE THREAD

**Issue by AlekseySh** [OML-Team/open-metric-learning#586]
We already have an example where a gallery set is fixed, but queries come online in batches: [link](https://github.com/OML-Team/open-metric-learning?tab=readme-ov-file#retrieval-by-trained-model-streaming--txt2im).

It would be great to improve its performance with vector databases.
The reworked example may look somehow like:

```python
...

# gallery is huge and fixed, so we only process it once
dataset_gallery = ImageBaseDataset(galleries, transform=transform)
embeddings_gallery = inference(extractor, dataset_gallery, batch_size=4, num_workers=0)

# ONE OF:
index = SklearnKNNIndex(embeddings_gallery)  # a child of IVectorIndex
index = FaissIndex(embeddings_gallery)  # a child of IVectorIndex
index = QdrantIndex(embeddings_gallery)  # a child of IVectorIndex

for queries in [queries1, queries2]:
    dataset_query = ImageBaseDataset(queries, transform=transform)
    embeddings_query = inference(extractor, dataset_query, batch_size=4, num_workers=0)

    # for the operation below we are going to provide integrations with vector search DB like QDrant or Faiss
    rr = RetrievalResults.from_index(
        index = index, embeddings_query=embeddings_query,
        dataset_query=dataset_query, dataset_gallery=dataset_gallery
    )
    rr = ConstantThresholding(th=80).process(rr)
    rr.visualize_qg([0, 1], dataset_query=dataset_query, dataset_gallery=dataset_gallery, show=True)
    print(rr)

```

I think we should start here with understanding of what `IVectorIndex` interface should include so it can handle different backends.

**Comment by Baduc81** [comment]
Hi, I'd like to work on this

Here's my proposed design for the `IVectorIndex` interface:

```python
class IVectorIndex(ABC):
    def add(self, embeddings: FloatTensor) -> None: ...
    def search(self, embeddings: FloatTensor, top_n: int) -> Tuple[FloatTensor, LongTensor]: ...
    def __len__(self) -> int: ...
```

`search()` returns `(distances, indices)` — L2 distances sorted ascending, matching the convention used elsewhere in OML.

I plan to implement three backends:
- `SklearnKNNIndex` — exact search via `sklearn.neighbors.NearestNeighbors`, no extra deps
- `FaissIndex` — exact/approximate search via `faiss.IndexFlatL2`, easily swappable to IVF/HNSW
- `QdrantIndex` — in-memory Qdrant collection, no server required

And add `RetrievalResults.from_index()` as shown in the issue description.

A few questions before I open a PR:
1. Should the backends be optional deps (guarded by `try/except ImportError`) or added to `requirements.txt`?
2. Is there a preferred location for the new file — `oml/retrieval/vector_index.py`?
3. Any preference on test coverage (unit tests only, or also an integration example)?

Happy to adjust the design based on your feedback
